mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraApplicableFlagDAO.java

```diff
@@ -0,0 +1,112 @@
+
+import static com.datastax.driver.core.querybuilder.QueryBuilder.add;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.bindMarker;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.eq;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.select;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.set;
+import static com.datastax.driver.core.querybuilder.QueryBuilder.update;
+import static org.apache.james.mailbox.cassandra.table.CassandraApplicableFlagTable.FIELDS;
+import static org.apache.james.mailbox.cassandra.table.CassandraApplicableFlagTable.MAILBOX_ID;
+import static org.apache.james.mailbox.cassandra.table.CassandraApplicableFlagTable.TABLE_NAME;
+import static org.apache.james.mailbox.cassandra.table.Flag.MAIL_FLAG_STRING;
+import static org.apache.james.mailbox.cassandra.table.Flag.USER_FLAGS;
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import javax.inject.Inject;
+import javax.mail.Flags;
+import javax.mail.Flags.Flag;
+
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.mailbox.FlagsBuilder;
+import org.apache.james.mailbox.cassandra.CassandraId;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.querybuilder.Update;
+import com.datastax.driver.core.querybuilder.Update.Assignments;
+
+public class CassandraApplicableFlagDAO {
+
+    private static final Flags.Flag[] ALL_APPLICABLE_FLAGS = {Flag.ANSWERED, Flag.DELETED, Flag.DRAFT, Flag.SEEN, Flag.FLAGGED };
+
+    public static final Flags EMPTY_FLAGS = new Flags();
+
+    private final CassandraAsyncExecutor cassandraAsyncExecutor;
+    private final PreparedStatement select;
+
+    @Inject
+    public CassandraApplicableFlagDAO(Session session) {
+        this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+        this.select = prepareSelect(session);
+    }
+
+    private PreparedStatement prepareSelect(Session session) {
+        return session.prepare(select(FIELDS)
+            .from(TABLE_NAME)
+            .where(eq(MAILBOX_ID, bindMarker(MAILBOX_ID))));
+    }
+
+    public CompletableFuture<Optional<Flags>> retrieveApplicableFlag(CassandraId mailboxId) {
+        return cassandraAsyncExecutor.executeSingleRow(
+            select.bind()
+                .setUUID(MAILBOX_ID, mailboxId.asUuid()))
+            .thenApply(rowOptional ->
+                rowOptional.map(row -> new FlagsExtractor(row).getApplicableFlags()));
+    }
+
+    public CompletableFuture<Void> updateApplicableFlags(CassandraId cassandraId, Flags flags) {
+        Flags newFlags = new FlagsBuilder().add(flags).build();
+        newFlags.remove(Flag.RECENT);
+        newFlags.remove(Flag.USER);
+
+        if (newFlags.equals(EMPTY_FLAGS)) {
+            return CompletableFuture.completedFuture(null);
```

Comment:
ApplicableFlags is the UNION of all flags that have been stored in the mailbox (set by the user)